flang/lib/Lower/OpenMP.cpp

```diff
@@ -142,6 +142,1122 @@ static void genNestedEvaluations(Fortran::lower::AbstractConverter &converter,
     converter.genEval(e);
 }
 
+//===----------------------------------------------------------------------===//
+// Clauses
+//===----------------------------------------------------------------------===//
+
+namespace detail {
+template <typename C> //
+llvm::omp::Clause getClauseIdForClass(C &&) {
+  using namespace Fortran;
+  using A = llvm::remove_cvref_t<C>; // A is referenced in OMP.inc
+  // The code included below contains a sequence of checks like the following
+  // for each OpenMP clause
+  //   if constexpr (std::is_same_v<A, parser::OmpClause::AcqRel>)
+  //     return llvm::omp::Clause::OMPC_acq_rel;
+  //   [...]
+#define GEN_FLANG_CLAUSE_PARSER_KIND_MAP
+#include "llvm/Frontend/OpenMP/OMP.inc"
+}
+} // namespace detail
+
+static llvm::omp::Clause getClauseId(const Fortran::parser::OmpClause &clause) {
+  return std::visit([](auto &&s) { return detail::getClauseIdForClass(s); },
+                    clause.u);
+}
+
+namespace omp {
+using namespace Fortran;
+using SomeType = evaluate::SomeType;
+using SomeExpr = semantics::SomeExpr;
+using MaybeExpr = semantics::MaybeExpr;
+
+template <typename T> //
+using List = std::vector<T>;
```

Comment:
It used to be, but some classes that ended up in `List` were too large for SmallVector's default size estimation (200-something bytes), so I converted it to `std::vector`.
To elaborate: `SmallVector` takes two template arguments: type and initial size.  The initial size used to be mandatory, but at some point it was made optional, and the initial size was computed based on the type.  If the type exceeds some specific size, the estimation will no longer work, and the code will fail to compile.  Instead of adding a specific number, I just changed it to `std::vector`.  It can be changed back though.

Reply:
Ah yes, I'm aware of this limitation. Do you think it would be possible to store const references to those larger objects rather than full copies? I'm guessing these might be Symbol or Expression-related types, which I've seen trigger this sort of thing before. If we can't get around copying these big structures, then either keeping `std::vector` or `llvm::SmallVector<T, 0>` would be necessary.

Reply:
We're creating these objects, so we need to manage storage for them.  If we wanted to store references, we'd have to have another container that actually holds the referenced objects.

Reply:
I was hoping we only needed to create clause structures and we were able to keep references to original symbols and expressions, but we're creating new expressions here then so you're right that it doesn't make sense to try to store them as references. I think for cases like this `llvm::SmallVector<T, 0>` is preferred over `std::vector`, so I'd still suggest making that small change.

Reply:
Done